src/bitmap_font.c

```diff
@@ -0,0 +1,1387 @@
+    textObj->printedGlyphs = mem_heap_alloc_id(memID, maxTileRows * 16 * sizeof(u16));
+    textObj->printedGlyphCounts = mem_heap_alloc_id(memID, maxTileRows * 16 * sizeof(u8));
+    textObj->parseString = NULL;
+    textObj->parsedOutput = NULL;
+    bmp_font_obj_clear_print_data(textObj);
+
+    return textObj;
+}
+
+
+// Delete BitmapFontOBJ.
+void delete_bmp_font_obj(struct BitmapFontOBJ *textObj) {
+    mem_heap_dealloc(textObj->printedGlyphs);
+    mem_heap_dealloc(textObj->printedGlyphCounts);
+    if (textObj->parsedOutput != NULL) {
+        mem_heap_dealloc(textObj->parsedOutput);
+    }
+    mem_heap_dealloc(textObj);
+}
+
+
+// Set BitmapFontOBJ parseString() function and allocate space for parsedOutput.
+void bmp_font_obj_set_format_parser(struct BitmapFontOBJ *textObj, void *stringParserFunc, u32 maxOutputLength) {
+    textObj->parseString = stringParserFunc;
+    textObj->parsedOutput = mem_heap_alloc_id(textObj->memID, maxOutputLength);
+}
+
+
+// Set BitmapFontOBJ data.
+void bmp_font_obj_set_data(struct BitmapFontOBJ *textObj, const struct BitmapFontData *fonts, u16 baseTileNum, u8 maxTileRows, u32 unused1, u32 unused2, u16 *printedGlyphs, u8 *printedGlyphCounts) {
+    textObj->fonts = fonts;
+    textObj->baseTileNum = baseTileNum;
+    textObj->maxAllocatedTileRows = maxTileRows;
+    textObj->printedGlyphs = printedGlyphs;
+    textObj->printedGlyphCounts = printedGlyphCounts;
+    bmp_font_obj_clear_print_data(textObj);
+}
+
+
+// Clear BitmapFontOBJ printed glyph data.
+void bmp_font_obj_clear_print_data(struct BitmapFontOBJ *textObj) {
+    u32 i;
+
+    for (i = 0; i < (textObj->maxAllocatedTileRows * 16); i++) {
+        textObj->printedGlyphs[i] = 0;
+        textObj->printedGlyphCounts[i] = 0;
+    }
+}
+
+
+// Get total animation objects required for generated text (ignoring whitespace).
+u32 bmp_font_obj_get_anim_total(const char *string) {
+    u32 total = 0;
+
+    for (; string[0] != '\0'; string += 2) {
+        if (string[0] == '.') {
+            continue;
+        }
+
+        if (string[0] == ':') {
+            continue;
+        }
+
+        if (!bmp_font_obj_glyph_is_whitespace(string)) {
+            total++;
+        }
+    }
+
+    return total;
+}
+
+
+// Get glyph width.
+u32 bmp_font_obj_get_glyph_width(const struct BitmapFontData *font, const char *string) {
+    u32 latinSpacingWidth, glyphWidth;
+
+    if (bmp_font_obj_glyph_is_whitespace(string)) {
+        return font->whitespaceWidth;
+    }
+
+    if ((string[0] == '.') || (string[0] == ':')) {
+        return 0;
+    }
+
+    if (bmp_font_obj_get_latin_glyph_type(string) == F_TEXT_LATIN_HALFWIDTH) {
+        latinSpacingWidth = 8;
+        string = bmp_font_obj_convert_latin_hw_to_fw(&string[1]);
+    } else {
+        latinSpacingWidth = 0;
+    }
+
+    bmp_font_get_glyph(font, string, NULL, &glyphWidth);
+    return latinSpacingWidth + glyphWidth;
+}
+
+
+// Get font style/palette value.
+u8 bmp_font_obj_get_style_value(char c) {
```

Comment:
feel like this could more accurately be described as just a general function that gets the value of an ASCII hex digit